pkg/app/services/subscription/service.go

```diff
@@ -0,0 +1,354 @@
+			return nil, &apierrors.PendingError{}
+		}
+		if err := s.Cache.Set(key, cacheTTL, &idempotentRequest{Processing: true}); err != nil {
+			rc.Log.Warnf("Can't save sub idempotency to cache by key %s: %s", key, err)
+		}
+		defer func() {
+			err := s.Cache.Set(key, cacheTTL, &idempotentRequest{
+				Sub:        retSub,
+				Processing: false,
+			})
+			if err != nil {
+				rc.Log.Warnf("Can't save sub idempotency to cache by key %s: %s", key, err)
+			}
+		}()
+	}
+
+	sub := models.OrgSub{
+		OrgID:                   context.OrgID,
+		BillingUserID:           rc.User.ID,
+		PaymentGatewayCardToken: payload.PaymentGatewayCardToken,
+		SeatsCount:              payload.SeatsCount,
+		CommitState:             models.OrgSubCommitStateCreateInit,
+	}
+
+	var err error
+	if err = sub.Create(rc.DB); err != nil {
+		return nil, errors.Wrap(err, "can't create sub")
+	}
+
+	retSub, err = s.sendToCreateQueue(rc, &sub)
+	if err != nil {
+		return nil, err
+	}
+
+	rc.Log.Infof("Created sub %#v", retSub)
+	return retSub, nil
+}
+
+func (s basicService) sendToUpdateQueue(rc *request.AuthorizedContext, sub *models.OrgSub) error {
+	if err := s.UpdateQueue.Put(sub.ID); err != nil {
+		return errors.Wrap(err, "failed to put to create repos queue")
+	}
+	_, err := s.finishQueueSending(rc, sub, models.OrgSubCommitStateUpdateInit, models.OrgSubCommitStateUpdateSentToQueue)
+	return err
+}
+
+func (s *basicService) Update(rc *request.AuthorizedContext, context *request.OrgSubID, payload *SubPayload) error {
+	var org models.Org
+	if err := models.NewOrgQuerySet(rc.DB).IDEq(context.OrgID.OrgID).One(&org); err != nil {
+		return errors.Wrap(err, "failed to get org from db")
+	}
+	if err := s.isAdminCached(rc, &org); err != nil {
+		return errors.Wrap(err, "failed to check for admin")
+	}
+
+	var sub models.OrgSub
+	err := models.NewOrgSubQuerySet(rc.DB).IDEq(context.SubID.SubID).One(&sub)
+	if err == gorm.ErrRecordNotFound {
+		return apierrors.ErrNotFound
+	} else if err != nil {
+		return errors.Wrap(err, "failed to fetch scoped org sub")
+	}
+
+	if sub.IsUpdating() {
+		return errors.New("sub is already in process of updating")
+	}
+
+	query := models.NewOrgSubQuerySet(rc.DB).
+		IDEq(sub.ID).GetUpdater().
+		SetCommitState(models.OrgSubCommitStateUpdateInit)
+
+	if payload.PaymentGatewayCardToken != "" {
+		query = query.SetPaymentGatewayCardToken(payload.PaymentGatewayCardToken)
+	}
+	if payload.SeatsCount != 0 {
+		query = query.SetSeatsCount(payload.SeatsCount)
+	}
+
+	n, err := query.UpdateNum()
+	if err != nil {
+		return errors.Wrapf(err, "failed to update sub with id %d", sub.ID)
+	}
+
+	if n != 1 {
+		return errors.New("no rows were updated, this really shouldn't happen")
+	}
+
+	return s.sendToUpdateQueue(rc, &sub)
+}
+
+func (s *basicService) Delete(rc *request.AuthorizedContext, context *request.OrgSubID) error {
+	return errors.New("not implemented")
+}
+
+//TODO: This is a lot of code duplication between org and sub services,
```

Comment:
you can make class like `OrgAccessPolicy` implementing admin check, class instance can be a member of both services. It will avoid code duplication.
```go
type OrgAccessPolicy struct {
  pf providers.Factory
  cache cache.Cache
  log logiutil.Log
}

func (p *OrgAccessPolicy) CanRead(auth *models.Auth, org *models.Org) bool {}

func (p *OrgAccessPolicy) CanModify(auth *models.Auth, org *models.Org) bool {}
```